### DISCUSSION TITLE
enum bug?

### DISCUSSION THREAD

**Discussion by pbelbin** [prisma/prisma#2732] [Q&A] [answered]
In a pre-existing MySQL database, I have some tables with enum columns.

To research this a bit, I took the sample project schema and pushed created the equivalent by hand in MySQL, with a couple of additional columns to clarify the issue.

In one case, I have database enum values which start with numerics (which is not supported as an enum name), and, in other cases, I have multi word strings, with spaces separating the words.

Specifically: '24HR' and '12HR', and, 'really good' (for extra value).

Here is the result of introspecting the database:
```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "mysql"
  url      = "mysql://root:kit@localhost:3306/prisma1"
}

model post {
  author_id Int
  id        Int          @default(autoincrement()) @id
  published Int?         @default(0)
  rating    post_rating?
  title     String?
  user      user         @relation(fields: [author_id], references: [id])

  @@index([author_id], name: "author_ix")
}

model user {
  email  String?
  id     Int          @default(autoincrement()) @id
  name   String?
  rating user_rating?
  post   post[]
}

enum post_rating {
  GOOD
  BAD
  OTHER
}

enum user_rating {
  super
  duper
  good
  really_good @map("really good")
  meh
  HR @map("24HR")
  HR @map("12HR")
}
```

So, clearly there is a bug here, in that the generated enum names have created identical names for the enum values of '24HR' and '12HR', but it does appear that it's saying that it can map from the name to a value.

So, if I fix up the 2 'HR' enum property names to be HR12 and HR24 respectively, and then generate the client (TypeScript), and I look in the generated code, I never see 12HR or 24HR in the source code.  Only HR12 and HR24.

Here is the schema.prisma I ended up with:

If I run this client code:

```typescript
import { PrismaClient } from '@prisma/client'

const prisma = new PrismaClient()

// A `main` function so that you can use async/await
async function main() {
  const allUsers = await prisma.user.findMany();
  console.log(allUsers);
}

main()
  .catch(e => {
    throw e
  })
  .finally(async () => {
    await prisma.disconnect()
  })

```

I am presented with:

```
> ts-node ./script.ts

[
  {
    email: 'fred@flintstone.com',
    id: 1,
    name: 'fred flintstone',
    rating: 'super'
  },
  {
    email: 'barney@rubble.com',
    id: 2,
    name: 'barney rubble',
    rating: 'really_good'
  },
  {
    email: 'wilma@flintstone.com',
    id: 3,
    name: 'wilma flintstone',
    rating: 'HR24'
  }
]
```

For the enum fields, I only see the identifier that the schema.prisma shows, never the real value.

I suppose, since this is an 'object' level dump, it makes sense that the representation is showing the enum names, and not the underlying values.  ie: HR24 instead of 24HR which is actually in the database.

If I wrote code to 'export', I would have to provide a translation to actual text versus the object representation, and, if I wanted to 'import', I'd have again, to do a translation.  Or, is there a way to get the underlying value, and possibly even have the generated code do the translation too?

Or, should enums just be ignored, and treat them as String values, and deal with the list of valid values some other way?

I may be mistaken, but, I did not notice in the description of enums, any mention of the use of the `@map()`.

Aside from the introspection bug that produces duplicate enum identifiers (or, perhaps it's a feature that it does what it does, and doesn't just stop entirely), **what's the 'best practice' for using enums**?  Is it just to avoid enums, and consider them to be strings?  Or, is there a way to output/ingest the underlying values too (since the mapping between the 'db' and 'model' values are right there in the schema.prisma file)?

**Comment by ryands17** [comment, marked as answer]
Hey @pbelbin :wave: 

> So, if I fix up the 2 'HR' enum property names to be HR12 and HR24 respectively, and then generate the client (TypeScript), and I look in the generated code, I never see 12HR or 24HR in the source code. Only HR12 and HR24.

That's the way `@map` works. It maps the value in your `schema.prisma` to the one in the database (specified in the map parameter).

So by using the following:

```prisma
enum user_rating {
  super
  duper
  good
  really_good @map("really good")
  meh
  HR24 @map("24HR")
  HR12 @map("12HR")
}
```

The models generated will have `HR12` and `HR24` and while querying, Prisma will convert them to `24HR` and `12HR` respectively.

So to use them to add data, you would have to do something like this:

```ts
  await prisma.user.create({
    data: {
      name: 'abc',
      rating: 'HR12', // this will map to `12HR` in the database by Prisma
    },
  })
```

As for two HR's, this is an issue and I will create one for this. Prisma doesn't support identifiers starting with numbers and that would hold true for enums as well.

If you do not want to run into this problem, you could also use a relation table that maps these values as strings, and keep a join table for posts and ratings that map with id's.

**Reply by pbelbin** [reply]
Hey @ryands17 , thank you for the clarification.  Cool re @map and how it works.  

It would be good if the documentation was more clear about how the `@map` works! :)

Regarding the two `HR` enum values, thank you for opening the issue.

**Comment by ludralph** [comment]
Hi @pbelbin,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
